src/defck.rs

```diff
@@ -0,0 +1,461 @@
+            {
+                let theorem = self
+                    .nset
+                    .lookup_label(justif_label.value(buf))
+                    .unwrap()
+                    .atom;
+                let definition = self.nset.lookup_label(label.value(buf)).unwrap().atom;
+                self.result.justifications.insert(definition, theorem);
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    fn verify_definition_statement(
+        &mut self,
+        syntax_axiom: Label,
+        addr: StatementAddress,
+    ) -> Result<(), Diagnostic> {
+        let stmt = self.db.statement_by_address(addr);
+        let definition = self.nset.lookup_label(stmt.label()).unwrap().atom;
+
+        let syntax_addr = self.nset.lookup_label_by_atom(syntax_axiom).address;
+
+        // push the definition to the validated list early, so that later definitions
+        // aren't as messed up if this check fails
+        self.result.definitions.insert(definition);
+        if let Some(prev) = self.result.def_map.insert(syntax_axiom, definition) {
+            return Err(Diagnostic::DefCkDuplicateDefinition(
+                self.nset.atom_name(syntax_axiom).into(),
+                self.nset.lookup_label_by_atom(prev).address,
+            ));
+        }
+
+        if self.result.justifications.contains_key(&definition) {
+            // Skip definitional check for definitions having a justification.
+
+            // TODO check that the justification matches the definition
+        } else {
+            // Check that the top level of the definition is an equality
+            let Some(fmla) = self.stmts.get_formula(&stmt) else {
+                // Ok because the error would have been reported already
+                return Ok(())
+            };
+
+            let root = fmla.root(self.db);
+            let equality = root.label();
+            match self.equalities.get(&equality) {
+                None => {
+                    self.equalities.insert(equality, false); // suppress future errors
+                    return Err(Diagnostic::DefCkNotAnEquality(
+                        self.nset.atom_name(equality).into(),
+                        self.result
+                            .equalities_by_tc
+                            .iter()
+                            .map(|(&label, _)| self.nset.atom_name(label).into())
+                            .collect(),
+                    ));
+                }
+                Some(false) => return Ok(()), // already reported the error
+                Some(true) => {}
+            }
+
+            let Some(lhs) = root.nth_child(0) else {
+                return Err(Diagnostic::DefCkMalformedDefinition(syntax_addr));
+            };
+            if lhs.label() != syntax_axiom {
+                return Err(Diagnostic::DefCkMalformedDefinition(syntax_addr));
+            }
+
+            // TODO definition check
```

Comment:
Am I correct that the only missing check now, is to confirm that all variables are distinct?

Reply:
No, of mmj2's 6-point list I think we just have part of step 1. The most complicated part will be proving that dummy variables are bound.